### ISSUE TITLE
Android undefined symbol: slCreateEngine

### ISSUE THREAD

**Issue by hsdk123** [kcat/openal-soft#499]
I get errors along the lines of `undefined symbol: slCreateEngine` when building an app for Android.
It would be nice if the cmakelists could be updated to fetch the library if the backend is enabled.

**Comment by kcat** [comment]
It does:
https://github.com/kcat/openal-soft/blob/master/cmake/FindOpenSL.cmake#L49
```cmake
find_library(OPENSL_LIBRARY NAMES OpenSLES
             DOC "The OpenSL library"
)
```
https://github.com/kcat/openal-soft/blob/master/cmake/FindOpenSL.cmake#L59
```cmake
if(OPENSL_FOUND)
    set(OPENSL_LIBRARIES ${OPENSL_LIBRARY})
    set(OPENSL_INCLUDE_DIRS ${OPENSL_INCLUDE_DIR} ${OPENSL_ANDROID_INCLUDE_DIR})
endif()
```
https://github.com/kcat/openal-soft/blob/master/CMakeLists.txt#L1043
```cmake
find_package(OpenSL)
if(OPENSL_FOUND)
    option(ALSOFT_BACKEND_OPENSL "Enable OpenSL backend" ON)
    if(ALSOFT_BACKEND_OPENSL)
        set(HAVE_OPENSL 1)
        set(ALC_OBJS  ${ALC_OBJS} alc/backends/opensl.cpp alc/backends/opensl.h)
        set(BACKENDS  "${BACKENDS} OpenSL,")
        set(EXTRA_LIBS ${OPENSL_LIBRARIES} ${EXTRA_LIBS})
        set(INC_PATHS ${INC_PATHS} ${OPENSL_INCLUDE_DIRS})
    endif()
endif()
```
https://github.com/kcat/openal-soft/blob/master/CMakeLists.txt#L1281
```cmake
target_link_libraries(${IMPL_TARGET} PRIVATE common ${LINKER_FLAGS} ${EXTRA_LIBS} ${MATH_LIB})
```
So anything that links to `${IMPL_TARGET}` (`OpenAL` normally) should also automatically link with `OpenSLES`. How are you setting up your app to use OpenAL Soft? Did you build OpenAL Soft with any special options?

**Comment by MikuAuahDark** [comment]
I can reproduce this problem when you also compile the example program, mainly `altonegen` with this CMake invocation (command-line method documented by [Android NDK](https://developer.android.com/ndk/guides/cmake#command-line))

```
$ cmake -DCMAKE_TOOLCHAIN_FILE=$ANDROID_NDK_ROOT/build/cmake/android.toolchain.cmake -DANDROID_ABI=arm64-v8a -DANDROID_PLATFORM=21 -Bandroidoal/arm64-v8a -Hopenal-soft
```

On unrelated note, OpenAL-soft fails to find OpenSL if you invoke it like this instead (method documented by [cmake-toolchain](https://cmake.org/cmake/help/latest/manual/cmake-toolchains.7.html#cross-compiling-for-android))

```
$ cmake -DCMAKE_SYSTEM_NAME=Android -DCMAKE_SYSTEM_VERSION=21 -DCMAKE_ANDROID_ARCH_ABI=arm64-v8a -Bandroidoal/arm64-v8a -Hopenal-soft
```

**Comment by kcat** [comment]
I don't know what's going on here. The `-DANDROID_ABI=arm64-v8a` part seems to be what's breaking the link, but aside from using different paths for the libraries (`...sysroot/usr/lib/aarch64-linux-android...` instead of `...sysroot/usr/lib/arm-linux-androideabi...`) there doesn't seem to be anything different when linking. The listed `libOpenSLES.so` exists, and `liblog.so` is there with it. Changing the platform number doesn't help, while keeping it set to platform 21 but changing to `armeabi-v7a` fixes it. Funny enough, if I try to set `ANDROID_STL=c++_shared` with `arm64-v8a`, it complains it can't find `libc++_shared.so`, even though that wasn't explicitly part of the link anywhere. It almost seems like there's a bug with the NDK's compiler or toolchain that the proper path for libraries isn't being set.

> On unrelated note, OpenAL-soft fails to find OpenSL if you invoke it like this instead (method documented by [cmake-toolchain](https://cmake.org/cmake/help/latest/manual/cmake-toolchains.7.html#cross-compiling-for-android))

That should be part of a toolchain file. It will also need the proper compilers and flags for the paths to the headers and libs set somewhere/somehow. I don't know how it all really works, and just prefer to let the NDK's toolchain do all its magic.

**Comment by MikuAuahDark** [comment]
I think the ARMv7 toolchain is bit odd when it comes to unresolved symbols, so it probably doesn't really fixes it but "hides" it until the app is run. A simple library loading test would be required for this, probably with ADB or something.

I'd expect same thing as this MikuAuahDark/livesim3-love-android#5 (see the OP logcat) probably happends on ARMv7 build of OpenAL-soft, and if you read till the end, my fix was as simple as explicitly linking to Android `log` library.

That's being said, I probably test how this is going tomorrow, if I didn't forgot.

**Comment by kcat** [comment]
Hmm, if it's true that `armeabi-v7a` also doesn't find the libraries at build time but only complains if they're not found at runtime, whereas `arm64-v8a` does complain at build time, that would be odd. It links `libopenal.so` with `-llog` and doesn't complain about not finding `liblog.so`, but when it links `alrecord` with `libopenal.so`, it does complain about not finding `liblog.so` and `libOpenSLES.so`. Other than libopenal.so linking with `clang++` and `-fPIC`, and alrecord linking with `clang` and `-Wl,--gc-sections`, I see no differences in the command line that would cause one to find `liblog.so`/`libOpenSLES.so` and the other to not. Unless shared library linking for Android ignores whether shared library inputs actually exist, and just records them as needed.

I don't know what Android and the NDK is expecting here. CMake is being used correctly to link libraries and apps, but the apps are failing to link because it can't find its own libs.

Sure enough, a quick test with the two sources:
```c++
// testlib.cpp

#include <iostream>

extern "C" int myfunc()
{
    std::cout<< "Hello World!\n";
    return 0;
}
```
```c
// test.c

int myfunc();
int main()
{
    return myfunc();
}
```
```cmake
# CMakeLists.txt

project(mytest)

add_library(testlib SHARED testlib.cpp)
add_executable(mytest test.c)
target_link_libraries(mytest PRIVATE testlib)
```
If you try to build it with:
```sh
$ cmake -DCMAKE_TOOLCHAIN_FILE=${NDK_ROOT}/build/cmake/android.toolchain.cmake -DANDROID_ABI=arm64-v8a -DANDROID_STL=c++_shared .
[...]
$ make
```
The executable will fail to link because it can't find `libc++_shared.so` that `libtestlib.so` needs. Interestingly, if the application target uses C++ instead of C, there's no error about not finding `libc++_shared.so`. Unfortunately, that doesn't extend to the other libs; if I change alrecord to be C++, it still fails to link and not find `libOpenSLES.so` and `liblog.so` that `libopenal.so` wants (but doesn't complain about `libc++_shared.so` anymore). I'm completely stumped.

**Comment by MikuAuahDark** [comment]
Ah looks like android/ndk#1183 explains everything.

Anyway I tested compiling OpenAL-soft with CMake without that `ANDROID_LD=lld` option for ARMv7, push it to my phone and try to run the example `altonegen` program with `adb shell`. The result? surprisingly it runs and outputs sine wave as expected. Although I don't quite remember if that linking problem happends in phones running 64-bit or in later Android versions, but to save trouble in the future, using lld is sufficient.

**TL;DR**: Either use `lld` (`-fuse-ld=lld` when compiling with command line directly or `-DANDROID_LD=lld` when compiling with CMake) or don't compile the example and utilities programs.

**Comment by kcat** [comment]
Added a comment in commit 0c510416db6acbc9da3536d005f7719ea1324ed1. Feel free to say if you think it can be clearer.
